crates/runtime/src/datafusion/param_utils.rs

```diff
@@ -0,0 +1,269 @@
+use std::collections::HashMap;
+
+use datafusion::{common::ParamValues, scalar::ScalarValue};
+use serde_json::Value;
+
+/// Converts a serde_json::Value into a datafusion::common::ParamValues.
+///
+/// # Arguments
+/// * `json` - The JSON value to convert.
+///
+/// # Returns
+/// * `Result<ParamValues, String>` - The converted ParamValues or an error message.
+///
+/// # Supported JSON Formats
+/// - Array: Converted to positional parameters (ParamValues::Positional).
+/// - Object: Converted to named parameters (ParamValues::Named).
+/// - Primitive types (null, bool, number, string) are supported within arrays or objects.
+///
+/// # Errors
+/// - Returns an error if the top-level JSON is not an array or object.
+/// - Returns an error if any value cannot be converted to a ScalarValue.
+pub fn convert_json_to_param_values(json: Value) -> Result<ParamValues, String> {
```

Comment:
This should be a Snafu error type instead of a String error type. ref: https://github.com/spiceai/spiceai/blob/trunk/docs/dev/style_guide.md#all-errors-use-snafu-functionality